subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/EventFiringTaskExecuter.java

```diff
@@ -53,15 +56,25 @@ public TaskExecuterResult call(BuildOperationContext operationContext) {
             }
 
             private TaskExecuterResult executeTask(BuildOperationContext operationContext) {
+                Logger logger = task.getLogger();
+                ContextAwareTaskLogger contextAwareTaskLogger = null;
                 try {
                     taskExecutionListener.beforeExecute(task);
+                    BuildOperationRef currentOperation = buildOperationExecutor.getCurrentOperation();
+                    if (logger instanceof ContextAwareTaskLogger) {
+                        contextAwareTaskLogger = (ContextAwareTaskLogger) logger;
+                        contextAwareTaskLogger.setFallbackBuildOperationId(currentOperation.getId());
+                    }
                 } catch (Throwable t) {
                     state.setOutcome(new TaskExecutionException(task, t));
                     return TaskExecuterResult.WITHOUT_OUTPUTS;
                 }
 
                 TaskExecuterResult result = delegate.execute(task, state, context);
 
+                if (contextAwareTaskLogger != null) {
```

Comment:
Above line never throws?

Reply:
it throws when any build is using `TaskInternal#replaceLogger` and providing a custom task. This is nothing that we support and is currently only used by Kotlin dsl plugin that implements C ContextAwareTaskLogger. But since the api is there we should be better safe than sorry here IMO.

Reply:
I mean can `delegate.execute` ever throw, leaving us without having reset the fallback build op id? Ie do we need try / finally here?

Reply:
The protocol for `delegate.execute` is that it will not propagate exceptions during task executions. failing tasks will package related exceptions in the provided `TaskState`.

Reply:
Thanks